### PR TITLE
feat: implement `OnceLock`

### PR DESCRIPTION
**Pull request by chise0713**
This PR is based on the `OnceLock` implementation from the standard library, with modifications to adapt it to our `Once` implementation.

Related: #440

Tests may benefit from further validation and edge case coverage.

Any feedback on the design and implementation would be appreciated.

### CODE REVIEW

File: src/once.rs [hunk 1]
```diff
@@ -196,7 +226,12 @@ impl Once {
     // currently no way to take an `FnOnce` and call it via virtual dispatch
     // without some allocation overhead.
     #[cold]
-    fn call_once_slow(&self, ignore_poison: bool, f: &mut dyn FnMut(OnceState)) {
+    fn call_once_slow(
+        &self,
+        ignore_poison: bool,
+        from_once_lock: bool, // work around for `OnceLock::get_or_try_init`
```
Comment (chise0713) [inline src/once.rs]: To fix incorrect `Err` handling in `OnceLock::get_or_try_init`
and align with std semantics, I extended `Once`'s internal API
to allow manual control over its state.

This keeps the public API unchanged, but adds a `OnceLock`-specific
code path internally.

I'm not entirely confident this is the best approach, so I'd really
appreciate any feedback!

File: src/once_lock.rs [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+///         42
+///     })
+/// }
+/// ```
+pub struct OnceLock<T> {
+    once: Once,
+    value: UnsafeCell<MaybeUninit<T>>,
+}
+
+impl<T> OnceLock<T> {
```
Comment (chise0713) [inline src/once_lock.rs:43]: The documentation is mostly adapted from the standard library.
